Unify different ranges into 1-based closed range

r6eve opened the pull request:
Karkinos treats different ranges something like BED format (0-based half-opened range) and SAMRecord of htsjdk (1-based closed range), but doesn't convert those into specified range.
This PR enables karkinos to unify those intervals into 1-based closed range.

This changes includes the followings.
  - Unify internal ranges (`CapInterval`, `CopyNumberInterval`) into 1-based closed ranges
  - Fix length values calculated from `SAMRecord` to act like 1-based closed range
  - Upgrade gradlew
  - Add unit testing
  - Refactor some codes

r6eve committed:
Bump up gradlew version to 4.10.2

r6eve committed:
Prepare for unit tests

r6eve committed:
Remove unnecessary imports

r6eve committed:
Fix length of CapInterval and add unit tests

r6eve committed:
Fix interval range

r6eve committed:
Add tests for CopyNumberInterval and CapInterval

r6eve committed:
Refactor class

r6eve committed:
Remove unnecessary imports

r6eve committed:
Fix interval of SAMRecord

alumi commented on src/main/java/jp/ac/utokyo/rcast/karkinos/exec/CapInterval.java:
```diff
@@ -118,155 +148,113 @@ public float getCgParcent() {
 
+	// The arguments take closed range [s,e]
+	public boolean intersect(final int s, final int e) {
+		final boolean overlap = start <= e && s <= end;
```
```suggestion
		return start <= e && s <= end;
```

alumi commented on src/main/java/jp/ac/utokyo/rcast/karkinos/exec/CapInterval.java:
```diff
@@ -118,155 +148,113 @@ public float getCgParcent() {
-	public boolean include(int pos, int mergin) {
-		boolean overlap = (start-mergin <= pos && pos <= end+mergin);
+	public boolean include(final int pos, final int mergin) {
+		final boolean overlap = start - mergin <= pos && pos <= end + mergin;
```
```suggestion
		return start - mergin <= pos && pos <= end + mergin;
```

alumi commented on src/main/java/jp/ac/utokyo/rcast/karkinos/exec/CapInterval.java:
```diff
@@ -118,155 +148,113 @@ public float getCgParcent() {
 
-	public boolean include(int pos, int mergin) {
-		boolean overlap = (start-mergin <= pos && pos <= end+mergin);
+	public boolean include(final int pos, final int mergin) {
```
This method can be merged with `public boolean include(final int pos)` at L:249

alumi commented on src/main/java/jp/ac/utokyo/rcast/karkinos/exec/CapInterval.java:
```diff
@@ -275,39 +263,35 @@ public double getCN() {
+		return hmmvalue;
+	}
+
+	public int getTotallength() {
```
This method seems to be unused.

alumi commented on src/main/java/jp/ac/utokyo/rcast/karkinos/exec/CaptureHolder.java:
```diff
@@ -260,19 +264,21 @@ public void loadTargetBedFirstTime(File bed, TwoBitGenomeReader tgr)
 					int start = cp.getStart();
 					int end = cp.getEnd();
-					int length = end - start;
+					int length = end - start + 1;
```
Please add a comment describing why we should not use `cp.getLength()` here

alumi reviewed (commented):
Thank you for really nice improvements!

`CapInterval#getStart` is called by
- `CheckPossibleHDAmp` L195
- `CNVUtils` L110
- `TwoStateHMM` L77, L138, L150, L167, L190, L198, L258, L353, L365
- `BaitSampling` L75, L131
- `CaptureHolder` L240, L250, L271, L298
- `DataSet` L271
- `PileUp` L66, L83
- `FileOutPut` L417, L434
- `GeneEachCNV` L32

then

- `GeneEachCNV` calls `GeneExons#getGeneId`
- `FileOutPut` calls `GeneExons#getGeneIntervalExon`

`GeneExons` loads `refGene.txt` and stores `txStart` and `txEnd` values which are 0-based half-open. I think we need to check them, too.

https://github.com/genome-rcast/karkinos/blob/384514f68e3bcdce51adac36d47edd079d1922e5/src/main/java/jp/ac/utokyo/rcast/karkinos/readssummary/GeneExons.java#L102-L110

Also, I guess we need to fix FileOutPut L417 to keep the output file format.

https://github.com/genome-rcast/karkinos/blob/384514f68e3bcdce51adac36d47edd079d1922e5/src/main/java/jp/ac/utokyo/rcast/karkinos/graph/output/FileOutPut.java#L417

r6eve committed:
Remove meaningless variables and method

r6eve committed:
Merge two methods into one

r6eve committed:
Add a comment to prevent from using invalid method

r6eve committed:
Convert the range of internal refGene from 0-based half-opened into 1-based closed-open

r6eve committed:
Fix Interval

r6eve commented:
Thank you for pointing out my mistake. I've fixed it accordingly.

athos commented on test/main/java/jp/ac/utokyo/rcast/karkinos/alleliccnv/TwoStateHMMTest.java:
```diff
@@ -0,0 +1,123 @@
+    CopyNumberInterval tmp_cni = new CopyNumberInterval();
+    tmp_cni.setStart(4);
+    tmp_cni.setEnd(6);
+    final CopyNumberInterval cni = tmp_cni;
```
Seems no need to have `tmp_cni` here. This will probably suffice:

```java
final CopyNumberInterval cni = new CopyNumberInterval();
cni.setStart(4);
cni.setEnd(6);
```

athos commented on test/main/java/jp/ac/utokyo/rcast/karkinos/exec/CapIntervalTest.java:
```diff
@@ -0,0 +1,122 @@
+      assertEquals(expected[i], ci.intersect(record));
+      ++i;
+    }
+    it.close();
```
This could be a more idiomatic Java-ish code, something like:

```java
try (SAMFileReader rdr = new SAMFileReader(file)) {
  final boolean[] expected = new boolean[] {true, true, true, false, false, false};
  int i = 0;
  for (SAMRecord record : rdr) {
    assertEquals(expected[i], ci.intersect(record));
    ++i;
  }
}
```

Also, you may want to use `SamReader` + `SamReaderFactory` instead of `SAMFileReader`, which is already a deprecated API.

r6eve replied:
Thank you for your advice. I updated the following codes.

```diff
-try (final SAMFileReader rdr = new SAMFileReader(file)) {
-  final boolean[] expected = new boolean[] {true, true, true, false, false, false};
-  int i = 0; 
-  for (final SAMRecord record : rdr) {
-    assertEquals(expected[i], ci.intersect(record));
-    ++i;
-  }
-}
+try (final SamReader rdr = SamReaderFactory.make().open(file)) {
+  final boolean[] expected = new boolean[] {true, true, true, false, false, false};
+  int i = 0; 
+  for (final SAMRecord record : rdr) {
+    assertEquals(expected[i], ci.intersect(record));
+    ++i;
+  }
+} catch (final Exception e) {
+  e.printStackTrace();
+}
```

Then, I got the following warning messages.

```
jp.ac.utokyo.rcast.karkinos.exec.CapIntervalTest > intersectFromSAMFileTest() STANDARD_ERROR
    DEBUG       2018-12-05 10:51:16     SRAAccession    Checking if SRA module is supported in that environment
    java.io.IOException: Server returned HTTP response code: 403 for URL: http://trace.ncbi.nlm.nih.gov/Traces/sratoolkit/sratoolkit.cgi
    java.io.IOException: Server returned HTTP response code: 403 for URL: http://trace.ncbi.nlm.nih.gov/Traces/sratoolkit/sratoolkit.cgi
    java.io.IOException: Server returned HTTP response code: 403 for URL: http://trace.ncbi.nlm.nih.gov/Traces/sratoolkit/sratoolkit.cgi
    Failed to load ngs-sdk library
    java.io.IOException: Server returned HTTP response code: 403 for URL: http://trace.ncbi.nlm.nih.gov/Traces/sratoolkit/sratoolkit.cgi
    java.io.IOException: Server returned HTTP response code: 403 for URL: http://trace.ncbi.nlm.nih.gov/Traces/sratoolkit/sratoolkit.cgi
    Cannot download library: -2
    Failed to download ngs-sdk from NCBI
    INFO        2018-12-05 10:51:19     SRAAccession    SRA is not supported. Will not be able to read from SRA
```

As far as I know, we cannot prohibit `open#SamReaderFactory` from trying to download.
So, I chose `SAMFileReader` in current version.

athos replied:
Looks weird.  `SamReaderFactor` seems to work fine on my environment even without internet connection.

But, that isn't worth digging deeper into it now, so I think choosing `SAMFileReader` is fair enough. Thanks for the change, and for taking the time to try to resolve the issue!

athos commented on test/main/java/jp/ac/utokyo/rcast/karkinos/alleliccnv/CheckPossibleHDAmpTest.java:
```diff
@@ -0,0 +1,53 @@
+    CopyNumberInterval tmp_cni = new CopyNumberInterval();
+    tmp_cni.setStart(2 * unit + 4);
+    tmp_cni.setEnd(2 * unit + 6);
+    final CopyNumberInterval cni = tmp_cni;
```
This will probably suffice:

```java
final CopyNumberInterval cni = new CopyNumberInterval();
cni.setStart(2 * unit + 4);
cni.setEnd(2 * unit + 6);
```

athos reviewed (commented):
Thanks for your great steps toward codebase improvements!

The changes themselves basically look good to me 👍 I added some minor comments (on more of coding style matters).

r6eve committed:
Remove meaningless temporary variables

r6eve committed:
Use for-each loop

alumi commented:
@r6eve Thank you for the fixes! Please give me some more time to check if there're no degradations.

r6eve commented:
Sorry, I forgot to unify some points in `refGene.txt`. I'll make test and fix it.
https://github.com/genome-rcast/karkinos/blob/f6bca10f0821d4042d46c0f1cd971882d60f5380/src/main/java/jp/ac/utokyo/rcast/karkinos/readssummary/GeneExons.java#L114-L149

r6eve committed:
Add tests for GeneExons

r6eve commented:
I'm super sorry to confuse you, it's my over-thinking.
I've added testcases for `refGene.txt`.

r6eve committed:
Refactor method and add comments for future work

r6eve commented:
I've refactored method which loads `refGene.txt` and added comments for future work. Please review it again.

alumi commented on src/main/java/jp/ac/utokyo/rcast/karkinos/readssummary/GeneExons.java:
```diff
@@ -75,95 +86,72 @@ public void setTumortotal(long tumortotal) {
+
+	private void loadmap(final String refflat) {
+		try (final CSVReader brcvs = new CSVReader(new FileReader(refflat), '\t')) {
+			try {
```
We can remove this `try`-`catch`.

alumi commented on src/main/java/jp/ac/utokyo/rcast/karkinos/readssummary/GeneExons.java:
```diff
@@ -75,93 +86,72 @@ public void setTumortotal(long tumortotal) {
+					final int txStart = Integer.parseInt(data[4]);
+					final int txEnd = Integer.parseInt(data[5]);
+					this.genemap
+					    .computeIfAbsent(chrom, k -> new TreeMap<Integer, Interval>())
```
Please use the diamond operator.

```suggestion
					    .computeIfAbsent(chrom, k -> new TreeMap<>())
```

alumi commented on src/main/java/jp/ac/utokyo/rcast/karkinos/readssummary/GeneExons.java:
```diff
@@ -75,93 +86,72 @@ public void setTumortotal(long tumortotal) {
+					final List<Integer> exonStarts = toIntList(data[9]);
+					final List<Integer> exonEnds = toIntList(data[10]);
+
+					// XXX: Is this loop intentional?
```
I guess this is an existing bug. Values of `exonStarts` and `exonEnds` with the same indices should be paired to represent intervals.

c.f. https://github.com/chrovis/varity/blob/8864f8474faa175bcbe271b259ee70f3a50cd506/src/varity/ref_gene.clj#L52-L55

alumi commented on src/main/java/jp/ac/utokyo/rcast/karkinos/readssummary/GeneExons.java:
```diff
@@ -75,93 +86,72 @@ public void setTumortotal(long tumortotal) {
+					// into Interval format (1-based closed open range).
+					final String refseqid = data[1];
+					final String chrom = data[2];
+					final int txStart = Integer.parseInt(data[4]);
```
`+1` here because we don't need the original value.

```suggestion
					final int txStart = Integer.parseInt(data[4]) + 1;
```

alumi commented on src/main/java/jp/ac/utokyo/rcast/karkinos/readssummary/GeneExons.java:
```diff
@@ -75,93 +86,72 @@ public void setTumortotal(long tumortotal) {
+					final int txEnd = Integer.parseInt(data[5]);
+					this.genemap
+					    .computeIfAbsent(chrom, k -> new TreeMap<Integer, Interval>())
+					    .put(txStart + 1, new Interval(chrom, txStart + 1, txEnd, refseqid, geneSymbol));
```
```suggestion
					    .put(txStart, new Interval(chrom, txStart, txEnd, refseqid, geneSymbol));
```

alumi commented on src/main/java/jp/ac/utokyo/rcast/karkinos/readssummary/GeneExons.java:
```diff
@@ -75,93 +86,72 @@ public void setTumortotal(long tumortotal) {
+					    .computeIfAbsent(chrom, k -> new TreeMap<Integer, Interval>())
+					    .put(txStart + 1, new Interval(chrom, txStart + 1, txEnd, refseqid, geneSymbol));
+
+					final int cdsStart = Integer.parseInt(data[6]);
```
```suggestion
					final int cdsStart = Integer.parseInt(data[6]) + 1;
```

alumi commented on src/main/java/jp/ac/utokyo/rcast/karkinos/readssummary/GeneExons.java:
```diff
@@ -75,93 +86,72 @@ public void setTumortotal(long tumortotal) {
-						tm.put(start, iv);
+						iv.exonidx = idx + 1;
+						this.map
+						    .computeIfAbsent(chrom, k -> new TreeMap<Integer, Interval>())
```
```suggestion
						    .computeIfAbsent(chrom, k -> new TreeMap<>())
```

alumi commented on src/main/java/jp/ac/utokyo/rcast/karkinos/readssummary/GeneExons.java:
```diff
@@ -75,93 +86,72 @@ public void setTumortotal(long tumortotal) {
+						int start = exonStarts.get(idx);
+						int end = exonEnds.get(idx);
+
+						// XXX: Is this condition intentional?
```
I think it's not. The condition can be simplified

c.f. https://github.com/chrovis/varity/blob/8864f8474faa175bcbe271b259ee70f3a50cd506/src/varity/ref_gene.clj#L172-L185

alumi commented on src/main/java/jp/ac/utokyo/rcast/karkinos/readssummary/GeneExons.java:
```diff
@@ -59,6 +59,17 @@ public long getNormaltotal() {
 		return normaltotal;
 	}
 
+	public Map<String, Integer> getCounter() {
```
You shouldn't make these accessors public only for testing

alumi commented on src/main/java/jp/ac/utokyo/rcast/karkinos/readssummary/GeneExons.java:
```diff
@@ -75,93 +86,72 @@ public void setTumortotal(long tumortotal) {
-	
-	public String symbolWithoutNum(String s){
+
+	public static String symbolWithoutNum(String s){
```
```suggestion
	private static String symbolWithoutNum(String s){
```

alumi commented on src/main/java/jp/ac/utokyo/rcast/karkinos/readssummary/GeneExons.java:
```diff
@@ -75,93 +86,72 @@ public void setTumortotal(long tumortotal) {
+					final int cdsStart = Integer.parseInt(data[6]);
+					final int cdsEnd = Integer.parseInt(data[7]);
+					final int exonCount = Integer.parseInt(data[8]);
+					final List<Integer> exonStarts = toIntList(data[9]);
```
`exonStarts` values should be incremented here for consistency

alumi commented on src/main/java/jp/ac/utokyo/rcast/karkinos/readssummary/GeneExons.java:
```diff
@@ -75,93 +86,72 @@ public void setTumortotal(long tumortotal) {
+
+						// XXX: Is this intentional? If the `depth` value is not used, we should call another constructor.
+						// Cf. https://github.com/genome-rcast/karkinos/blob/384514f68e3bcdce51adac36d47edd079d1922e5/src/main/java/jp/ac/utokyo/rcast/karkinos/readssummary/Interval.java#L25-L30
+						final Interval iv = new Interval(chrom, start, end);
```
Please replace with the another ctor

alumi commented on src/test/java/jp/ac/utokyo/rcast/karkinos/readssummary/GeneExonsTest.java:
```diff
@@ -1,25 +1,25 @@
-    GeneExons ex = new GeneExons(ref);
-  }
-}
+// /*
```
Please remove this file

alumi commented on src/main/java/jp/ac/utokyo/rcast/karkinos/readssummary/GeneExons.java:
```diff
@@ -75,93 +86,72 @@ public void setTumortotal(long tumortotal) {
-
-		} catch (Exception e) {
-			// TODO Auto-generated catch block
+		} catch (final Exception e) {
```
We can remove this `catch` and add `throws IOException` to the method signature and the ctor.

alumi commented on src/main/java/jp/ac/utokyo/rcast/karkinos/readssummary/GeneExons.java:
```diff
@@ -75,93 +86,72 @@ public void setTumortotal(long tumortotal) {
+			try {
+				String[] data = null;
+				while ((data = brcvs.readNext()) != null) {
+					final String geneSymbol = data[0];
```
This is irrelevant to this PR, but `geneSymbol` is stored at 

- `data[0]` of `refFlat.txt`
- `data[12]` of `refGene.txt`

We should determine the format by checking that `data.length` is `11` or `16`

alumi reviewed (commented):
Thank you for commits! I added some comments, but most of them are trivial.

r6eve committed:
Remove meaningless try-catch block

r6eve committed:
Use diamond operator

r6eve committed:
Discard original values

r6eve committed:
Remove an unused file

r6eve committed:
Fix exons included in a coding region

r6eve committed:
Make constructor throw IOException

r6eve committed:
Remove public accessors and access fields

r6eve committed:
Modify private static method

r6eve committed:
Fix `geneSymbol` and `symbolWithoutNum` fields

r6eve commented:
Thank you for your detailed comments. I've fixed it.

r6eve committed:
Fix interval for subMap style

r6eve committed:
Add some tests to check intervals

r6eve commented:
Sorry, again. I forgot to check [subMap](https://docs.oracle.com/javase/8/docs/api/java/util/TreeMap.html#subMap-K-K-). I've fixed it and added the tests.

r6eve committed:
Fix updating the end point to extend interval

r6eve committed:
Add test codes to test readssummary/Interval

r6eve commented:
I've unified the interval range and fixed the bug that doesn't extend the `Interval#end` correctly.

r6eve committed:
Fix typo `s/Hdelation/Hdeletion/g`

r6eve committed:
Remove unused or meaningless methods and introduce `length` method

r6eve committed:
Fix length of CopyNumberIntarval

r6eve committed:
Fix broken test

r6eve commented:
I've introduced the `CopyNumberIntarva#length` method which calculates the length correctly.

r6eve committed:
Add unit tests for GeneExons.java

r6eve committed:
Refactor GeneExons.java

r6eve commented:
In previous implementation (i.e. current master branch), there is a bug that [counter.get(Object key)](https://github.com/genome-rcast/karkinos/blob/384514f68e3bcdce51adac36d47edd079d1922e5/src/main/java/jp/ac/utokyo/rcast/karkinos/readssummary/GeneExons.java#L293) always returns a constant value. This is because the `counter` always puts same key `""` which is returned by `symbolWithoutNum(data[0])`.

I confirmed this bug has already been fixed by 158226192ca8a36474b299e2837de8aac399dcb3 . And, I added unit tests to check it.

r6eve commented:
I think all points have been clarified. Please review it again.

alumi commented on test/main/java/jp/ac/utokyo/rcast/karkinos/exec/CaptureHolderTest.java:
```diff
@@ -0,0 +1,86 @@
+    final CaptureHolder ch = new CaptureHolder();
+    try {
+      ch.loadTargetBedFirstTime(bed, new TwoBitGenomeReader(twobit));
+    } catch (final Exception e) {
```
I think it's better not to catch exceptions thrown in tests unless they're intended and recoverable errors.

alumi commented on test/main/java/jp/ac/utokyo/rcast/karkinos/alleliccnv/TwoStateHMMTest.java:
```diff
@@ -0,0 +1,123 @@
+      SummaryStatistics expected = new SummaryStatistics();
+      expected.addValue(1.0);
+      expected.addValue(10.0);
+      assertTrue(ss.equals(expected));
```
`assertEquals` might provide a better report when it failed, especially with IDEs.

```suggestion
      assertEquals(expected, ss);
```

alumi reviewed (dismissed):
LGTM 👍👍

r6eve committed:
Don't catch meaningless exceptions thrown in test

r6eve committed:
Make test results more developer-friendly

r6eve commented:
Thank you for pointing out. I fixed it.

athos reviewed (approved):
Sorry for blocking this accidentally 🙇 Just approved now.

alumi commented:
Merged. Thank you for the great improvement!

r6eve commented:
I appreciate all your reviews. Thank you.
